crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,265 @@
+            .expect("Should be able to set the global subscriber.");
+    }
+
+    tracing::info!("Logging initialized with level: {log_level}");
+}
+
+/// We never log during tests as tests run in parallel within a single process,
+/// but you can only have 1 global subscriber per process.
+///
+/// If you are debugging a single test, you can override this to emit messages to stderr.
+///
+/// Note that if you override this and run multiple tests in parallel, then the call
+/// to `set_global_default()` will error causing a panic.
+fn is_test_client(client_info: Option<&ClientInfo>) -> bool {
+    client_info.map_or(false, |client_info| client_info.name == "AirTestClient")
+}
+
+fn resolve_log_level(log_level: Option<LogLevel>) -> LogLevel {
+    // Check log environment variable, this overrides any Client options
+    if let Some(log_level) = std::env::var(LOG_ENV_KEY)
+        .ok()
+        .and_then(|level| serde_json::from_value(serde_json::Value::String(level)).ok())
+    {
+        return log_level;
+    }
+
+    // Client specified log level through initialization parameters
+    if let Some(log_level) = log_level {
+        return log_level;
+    }
+
+    // Fallback
+    LogLevel::default()
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum LogLevel {
+    Error,
+    Warn,
+    #[default]
+    Info,
+    Debug,
+    Trace,
+}
+
+impl LogLevel {
+    fn tracing_level(self) -> tracing::Level {
+        match self {
+            Self::Error => tracing::Level::ERROR,
+            Self::Warn => tracing::Level::WARN,
+            Self::Info => tracing::Level::INFO,
+            Self::Debug => tracing::Level::DEBUG,
+            Self::Trace => tracing::Level::TRACE,
+        }
+    }
+}
+
+impl Display for LogLevel {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Error => f.write_str("Error"),
+            Self::Warn => f.write_str("Warn"),
+            Self::Info => f.write_str("Info"),
+            Self::Debug => f.write_str("Debug"),
+            Self::Trace => f.write_str("Trace"),
+        }
+    }
+}
+
+/// Filters out logs which have a log level lower than the level set by the client.
+struct LogLevelFilter {
+    filter: LogLevel,
+}
+
+impl<S> tracing_subscriber::layer::Filter<S> for LogLevelFilter {
+    fn enabled(
+        &self,
+        meta: &tracing::Metadata<'_>,
+        _: &tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        // This is a potential reason to use `air_` as the crate prefix,
+        // it would make it easy to set the `tracing_level()` for only air related crates
+        let filter = if meta.target().starts_with("air") || meta.target().starts_with("lsp") {
+            self.filter.tracing_level()
+        } else {
+            tracing::Level::INFO
+        };
+
+        meta.level() <= &filter
```

Comment:
Remember that tons of other crates use `tracing::info!()` and friends internally

This filter uses `target`, which is typically the module name, to determine the tracing level to filter with, and we never go below `Info` for 3rd party crates.

This is actually one reason it would be quite nice to prefix all of our crates with `air_`! Detection would be very simple.

Reply:
I think the filter implemented in ark is more flexible? https://github.com/posit-dev/ark/blob/3b8ff6112d14f1b78640e4db55c82754c0b362bf/crates/ark/src/logger.rs#L26-L43

Is there any reason not to use this approach?